File: gripql/python/gripql/query.py
from __future__ import absolute_import, print_function, unicode_literals

try:
    # attempt to load JSON parsing library that works faster
    from orjson import loads as jloads, dumps as jdumps
except ImportError:
    # fall back to standard JSON parsing library
    from json import loads as jloads, dumps as jdumps

import logging
import requests

from gripql.util import BaseConnection, Rate, raise_for_status


def _wrap_value(value, typ):
    wrapped = []
    if isinstance(value, list):
        if not all(isinstance(i, typ) for i in value):
            raise TypeError("expected all values in array to be a string")
        wrapped = value
    elif isinstance(value, typ):
        wrapped.append(value)
    elif value is None:
        pass
    else:
        raise TypeError("expected value to be a %s" % typ)
    return wrapped


def _wrap_str_value(value):
    return _wrap_value(value, str)


def _wrap_dict_value(value):
    return _wrap_value(value, dict)


class Query(BaseConnection):
    def __init__(self, url, graph, user=None, password=None, token=None, credential_file=None, resume=None):
        super(Query, self).__init__(url, user, password, token, credential_file)
        self.url = self.base_url + "/v1/graph/" + graph + "/query"
        self.graph = graph
        self.query = []
        self.resume = resume

    def __append(self, part):
        q = self.__class__(self.base_url, self.graph, self.user, self.password, self.token, self.credential_file, self.resume)
        q.query = self.query[:]
        q.query.append(part)
        return q

    def V(self, id=[]):
        """
        Start the query at a vertex.

        "id" is an ID or a list of vertex IDs to start from. Optional.
        """
        id = _wrap_str_value(id)
        return self.__append({"v": id})

    def E(self, id=[]):
        """
        Start the query at an edge.

        "id" is an ID to start from. Optional.
        """
        id = _wrap_str_value(id)
        return self.__append({"e": id})

    def in_(self, label=[]):
        """
        Follow an incoming edge to the source vertex.

        "label" is the label of the edge to follow.
        "label" can be a list.
        """
        label = _wrap_str_value(label)
        return self.__append({"in": label})

    def inV(self, label=[]):
        return self.in_(label)

    def out(self, label=[]):
        """
        Follow an outgoing edge to the destination vertex.

        "label" is the label of the edge to follow.
        "label" can be a list.
        """
        label = _wrap_str_value(label)
        return self.__append({"out": label})

    def outV(self, label=[]):
        return self.out(label)

    def both(self, label=[]):
        """
        Follow both incoming and outgoing edges to vertices.

        "label" is the label of the edge to follow.
        "label" can be a list.
        """
        label = _wrap_str_value(label)
        return self.__append({"both": label})

    def bothV(self, label=[]):
        return self.both(label)

    def inE(self, label=[]):
        """
        Move from a vertex to an incoming edge.

        "label" is the label of the edge to move to.
        "label" can be a list.

        Must be called from a vertex.
        """
        label = _wrap_str_value(label)
        return self.__append({"inE": label})

    def outE(self, label=[]):
        """
        Move from a vertex to an outgoing edge.

        "label" is the label of the edge to move to.
        "label" can be a list.

        Must be called from a vertex.
        """
        label = _wrap_str_value(label)
        return self.__append({"outE": label})

    def bothE(self, label=[]):
        """
        Move from a vertex to incoming/outgoing edges.

        "label" is the label of the edge to move to.
        "label" can be a list.

        Must be called from a vertex.
        """
        label = _wrap_str_value(label)
        return self.__append({"bothE": label})

    def has(self, expression):
        """
        Filter vertex/edge based on properties.
        """
        return self.__append({"has": expression})

    def hasLabel(self, label):
        """
        Filter vertex/edge based on label.
        """
        label = _wrap_str_value(label)
        return self.__append({"hasLabel": label})

    def hasId(self, id):
        """
        Filter vertex/edge based on id.
        """
        id = _wrap_str_value(id)
        return self.__append({"hasId": id})

    def hasKey(self, key):
        """
        Filter vertex/edge based on the existence of properties.
        """
        key = _wrap_str_value(key)
        return self.__append({"hasKey": key})

    def fields(self, field=[]):
        """
        Select document properties to be returned in document.
        """
        field = _wrap_str_value(field)
        return self.__append({"fields": field})

    def as_(self, name):
        """
        Mark the current vertex/edge with the given name.

        Used to return elements from select().
        """
        return self.__append({"as": name})

    def select(self, marks):
        """
        Returns rows of marked elements, with one item for each mark.

        "marks" is a list of mark names.
        The rows returned are all combinations of marks, e.g.
        [
            [A1, B1],
            [A1, B2],
            [A2, B1],
            [A2, B2],
        ]
        """
        marks = _wrap_str_value(marks)
        return self.__append({"select": {"marks": marks}})

    def limit(self, n):
        """
        Limits the number of results returned.
        """
        return self.__append({"limit": n})

    def skip(self, n):
        """
        Offset the results returned.
        """
        return self.__append({"skip": n})

    def range(self, start, stop):
        """
        When the low-end of the range is not met, objects are continued to be iterated.
        When within the low (inclusive) and high (exclusive) range, traversers are emitted.
        When above the high range, the traversal breaks out of iteration. Finally, the use
        of -1 on the high range will emit remaining traversers after the low range begins.
        """
        return self.__append({"range": {"start": start, "stop": stop}})

    def count(self):
        """
        Return the number of results, instead of the elements.
        """
        return self.__append({"count": ""})

    def distinct(self, props=[]):
        """
        Select distinct elements based on the provided property list.
        """
        props = _wrap_str_value(props)
        return self.__append({"distinct": props})

    def render(self, template):
        """
        Render output of query
        """
        return self.__append({"render": template})

    def path(self):
        """
        Display path of query
        """
        return self.__append({"path": []})

    def unwind(self, field):
        """
        Unwind an array
        """
        return self.__append({"unwind": field})

    def aggregate(self, aggregations):
        """
        Aggregate results of query output
        """
        aggregations = _wrap_dict_value(aggregations)
        return self.__append({"aggregate": {"aggregations": aggregations}})

    def to_json(self):
        """
        Return the query as a JSON string.
        """
        output = {"query": self.query}
        return jdumps(output)

    def to_dict(self):
        """
        Return the query as a dictionary.
        """
        return {"query": self.query}

    def __iter__(self):
        return self.__stream()

    def __stream(self, raw=False, debug=False):
        """
        Execute the query and return an iterator.
        """
        log_level = logging.root.level
        if debug:
            log_level = logging.DEBUG
        logger = logging.getLogger(__name__)
        logger.handlers = []
        stream_handler = logging.StreamHandler()
        stream_handler.setFormatter(
            logging.Formatter('[%(levelname)s]\t%(asctime)s\t%(message)s')
        )
        stream_handler.setLevel(log_level)
        logger.setLevel(log_level)
        logger.addHandler(stream_handler)

        rate = Rate(logger)
        rate.init()
        if self.resume is None:
            response = self.session.post(
                self.url,
                json=self.to_dict(),
                stream=True
            )
            logger.debug('POST %s', self.url)
        else:
            url = self.base_url + "/v1/graph/" + self.graph + "/job-resume"
            data = self.to_dict()
            data['srcId'] = self.resume
            response = self.session.post(
                url,
                json=data,
                stream=True
            )
            logger.debug('POST %s', url)
        logger.debug('BODY %s', self.to_json())
        logger.debug('STATUS CODE %s', response.status_code)

        for result in response.iter_lines(chunk_size=None):
            try:
                result_dict = jloads(result.decode())
            except Exception as e:
                logger.error("Failed to decode: %s", result)
                raise e

            if raw:
                extracted = result_dict
            elif "vertex" in result_dict:
                extracted = result_dict["vertex"]
            elif "edge" in result_dict:
                extracted = result_dict["edge"]
            elif "aggregations" in result_dict:
                extracted = result_dict["aggregations"]
            elif "selections" in result_dict:
                extracted = result_dict["selections"]["selections"]
                for k in extracted:
                    if "vertex" in extracted[k]:
                        extracted[k] = extracted[k]["vertex"]
                    elif "edge" in extracted[k]:
                        extracted[k] = extracted[k]["edge"]
            elif "render" in result_dict:
                extracted = result_dict["render"]
            elif "path" in result_dict:
                extracted = result_dict["path"]
            elif "count" in result_dict:
                extracted = result_dict
            elif "error" in result_dict:
                raise requests.HTTPError(result_dict['error']['message'])
            else:
                extracted = result_dict

            yield extracted

            rate.tick()
        rate.close()

    def execute(self, stream=False, raw=False, debug=False):
        """
        Execute the query.

        If stream is True an iterator will be returned. Otherwise, a list
        is returned.
        """
        if stream:
            return self.__stream(raw=raw, debug=debug)
        else:
            output = []
            for r in self.__stream(raw=raw, debug=debug):
                output.append(r)
            return output

    def submit(self, debug=False):
        """
        Post the traversal as an asynchronous job
        """
        log_level = logging.root.level
        if debug:
            log_level = logging.DEBUG
        logger = logging.getLogger(__name__)
        logger.handlers = []
        stream_handler = logging.StreamHandler()
        stream_handler.setFormatter(
            logging.Formatter('[%(levelname)s]\t%(asctime)s\t%(message)s')
        )
        stream_handler.setLevel(log_level)
        logger.setLevel(log_level)
        logger.addHandler(stream_handler)

        url = self.base_url + "/v1/graph/" + self.graph + "/job"

        response = self.session.post(
            url,
            json=self.to_dict()
        )
        raise_for_status(response)
        return response.json()

    def searchJobs(self, debug=False):
        """
        Find jobs that match this query
        """
        log_level = logging.root.level
        if debug:
            log_level = logging.DEBUG
        logger = logging.getLogger(__name__)
        logger.handlers = []
        stream_handler = logging.StreamHandler()
        stream_handler.setFormatter(
            logging.Formatter('[%(levelname)s]\t%(asctime)s\t%(message)s')
        )
        stream_handler.setLevel(log_level)
        logger.setLevel(log_level)
        logger.addHandler(stream_handler)

        url = self.base_url + "/v1/graph/" + self.graph + "/job-search"

        response = self.session.post(
            url,
            json=self.to_dict()
        )
        for result in response.iter_lines(chunk_size=None):
            result_dict = jloads(result.decode())
            yield result_dict



class __Query(Query):
    def __init__(self):
        self.query = []

    def __append(self, part):
        q = self.__class__()
        q.query = self.query[:]
        q.query.append(part)
        return q

    def __iter__(self):
        pass

    def __stream(self):
        pass

    def execute(self):
        pass


__ = __Query()

File: gripql/query.go
package gripql

import (
	"fmt"
	"strings"

	"github.com/bmeg/grip/protoutil"
)

// V starts a new vertex query, short for `NewQuery().V()`.
func V(ids ...string) *Query {
	return NewQuery().V(ids...)
}

// E starts a new vertex query, short for `NewQuery().E()`.
func E(ids ...string) *Query {
	return NewQuery().E(ids...)
}

// Index starts a new vertex query using an index hits to start
func Index(field, term string) *Query {
	return NewQuery().Index(field,term)
}

// NewQuery creates a new Query instance.
func NewQuery() *Query {
	return &Query{}
}

// Query helps build graph queries.
type Query struct {
	Statements []*GraphStatement
}

func (q *Query) with(st *GraphStatement) *Query {
	nq := &Query{
		Statements: make([]*GraphStatement, len(q.Statements)),
	}
	copy(nq.Statements, q.Statements)
	nq.Statements = append(nq.Statements, st)
	return nq
}

// V adds a vertex selection step to the query
func (q *Query) V(id ...string) *Query {
	vlist := protoutil.AsListValue(id)
	return q.with(&GraphStatement{Statement: &GraphStatement_V{vlist}})
}

// E adds a edge selection step to the query
func (q *Query) E(id ...string) *Query {
	elist := protoutil.AsListValue(id)
	return q.with(&GraphStatement{Statement: &GraphStatement_E{elist}})
}

// Index adds a index selection step to the query
func (q *Query) Index(field, term string) *Query {
	return q.with(&GraphStatement{
		Statement: &GraphStatement_Index{ &IndexQuery{Field:field, Value:term},
	}})
}

// In follows incoming edges to adjacent vertex
func (q *Query) In(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_In{vlist}})
}

// InV follows incoming edges to adjacent vertex
func (q *Query) InV(label ...string) *Query {
	return q.In(label...)
}

// InE moves to incoming edge
func (q *Query) InE(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_InE{vlist}})
}

// Out follows outgoing edges to adjacent vertex
func (q *Query) Out(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_Out{vlist}})
}

// OutV follows outgoing edges to adjacent vertex
func (q *Query) OutV(label ...string) *Query {
	return q.Out(label...)
}

// OutE moves to outgoing edge
func (q *Query) OutE(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_OutE{vlist}})
}

// Both follows both incoming and outgoing edges to adjacent vertex
func (q *Query) Both(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_Both{vlist}})
}

// BothV follows both incoming and outgoing edges to adjacent vertex
func (q *Query) BothV(label ...string) *Query {
	return q.Both(label...)
}

// BothE moves to both incoming and outgoing edges
func (q *Query) BothE(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_BothE{vlist}})
}

// Has filters elements based on data properties.
func (q *Query) Has(expression *HasExpression) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Has{expression}})
}

// HasLabel filters elements based on their label.
func (q *Query) HasLabel(label ...string) *Query {
	vlist := protoutil.AsListValue(label)
	return q.with(&GraphStatement{Statement: &GraphStatement_HasLabel{vlist}})
}

// HasKey filters elements based on whether it has one or more properties.
func (q *Query) HasKey(key ...string) *Query {
	vlist := protoutil.AsListValue(key)
	return q.with(&GraphStatement{Statement: &GraphStatement_HasKey{vlist}})
}

// HasID filters elements based on their id.
func (q *Query) HasID(id ...string) *Query {
	vlist := protoutil.AsListValue(id)
	return q.with(&GraphStatement{Statement: &GraphStatement_HasId{vlist}})
}

// Limit limits the number of results returned.
func (q *Query) Limit(n uint32) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Limit{n}})
}

// Skip will drop the first n number of records and return the rest.
func (q *Query) Skip(n uint32) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Skip{n}})
}

// Range will limits which records are returned. When the low-end of the range is
// not met, objects are continued to be iterated. When within the low (inclusive)
// and high (exclusive) range, traversers are emitted. When above the high range,
// the traversal breaks out of iteration. Finally, the use of -1 on the high range
// will emit remaining traversers after the low range begins.
func (q *Query) Range(start, stop int32) *Query {
	return q.with(&GraphStatement{
		Statement: &GraphStatement_Range{
			&Range{
				Start: start,
				Stop:  stop,
			},
		},
	})
}

// As marks current elements with tag
func (q *Query) As(id string) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_As{id}})
}

// Select retreieves previously marked elemets
func (q *Query) Select(id ...string) *Query {
	idList := SelectStatement{Marks: id}
	return q.with(&GraphStatement{Statement: &GraphStatement_Select{&idList}})
}

// Fields selects which properties are returned in the result.
func (q *Query) Fields(keys ...string) *Query {
	klist := protoutil.AsListValue(keys)
	return q.with(&GraphStatement{Statement: &GraphStatement_Fields{klist}})
}

// Match is used to concatenate multiple queries.
func (q *Query) Match(qs ...*Query) *Query {
	queries := []*QuerySet{}
	for _, q := range qs {
		queries = append(queries, &QuerySet{
			Query: q.Statements,
		})
	}
	set := &MatchQuerySet{Queries: queries}
	return q.with(&GraphStatement{Statement: &GraphStatement_Match{set}})
}

// Count adds a count step to the query
func (q *Query) Count() *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Count{}})
}

// Distinct selects records with distinct elements of arg
func (q *Query) Distinct(args ...string) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Distinct{protoutil.AsListValue(args)}})
}

// Render adds a render step to the query
func (q *Query) Render(template interface{}) *Query {
	return q.with(&GraphStatement{Statement: &GraphStatement_Render{protoutil.WrapValue(template)}})
}

func (q *Query) String() string {
	parts := []string{}
	add := func(name string, x ...string) {
		args := strings.Join(x, ", ")
		parts = append(parts, fmt.Sprintf("%s(%s)", name, args))
	}

	for _, gs := range q.Statements {
		switch stmt := gs.GetStatement().(type) {
		case *GraphStatement_V:
			ids := protoutil.AsStringList(stmt.V)
			add("V", ids...)

		case *GraphStatement_E:
			ids := protoutil.AsStringList(stmt.E)
			add("E", ids...)

		case *GraphStatement_In:
			ids := protoutil.AsStringList(stmt.In)
			add("In", ids...)

		case *GraphStatement_Out:
			ids := protoutil.AsStringList(stmt.Out)
			add("Out", ids...)

		case *GraphStatement_Both:
			ids := protoutil.AsStringList(stmt.Both)
			add("Both", ids...)

		case *GraphStatement_InE:
			ids := protoutil.AsStringList(stmt.InE)
			add("InE", ids...)

		case *GraphStatement_OutE:
			ids := protoutil.AsStringList(stmt.OutE)
			add("OutE", ids...)

		case *GraphStatement_BothE:
			ids := protoutil.AsStringList(stmt.BothE)
			add("BothE", ids...)

		case *GraphStatement_Has:
			add("Has")

		case *GraphStatement_HasLabel:
			labels := protoutil.AsStringList(stmt.HasLabel)
			add("HasLabel", labels...)

		case *GraphStatement_HasId:
			ids := protoutil.AsStringList(stmt.HasId)
			add("HasId", ids...)

		case *GraphStatement_HasKey:
			keys := protoutil.AsStringList(stmt.HasKey)
			add("HasKey", keys...)

		case *GraphStatement_Limit:
			add("Limit", fmt.Sprintf("%d", stmt.Limit))

		case *GraphStatement_Skip:
			add("Skip", fmt.Sprintf("%d", stmt.Skip))

		case *GraphStatement_Range:
			add("Range", fmt.Sprintf("%d", stmt.Range.Start), fmt.Sprintf("%d", stmt.Range.Stop))

		case *GraphStatement_Count:
			add("Count")

		case *GraphStatement_As:
			add("As", stmt.As)

		case *GraphStatement_Select:
			add("Select", stmt.Select.Marks...)

		case *GraphStatement_Match:
			add("Match")

		case *GraphStatement_Fields:
			fields := protoutil.AsStringList(stmt.Fields)
			add("Fields", fields...)

		case *GraphStatement_Aggregate:
			add("Aggregate")

		case *GraphStatement_Render:
			add("Render")
		}
	}

	return strings.Join(parts, ".")
}
